cirrus/google_cloud/utils.py

```diff
@@ -230,6 +236,126 @@ def _get_string_to_sign(
+            everything that follows the host name but precedes any query strings.
+        http_verb (str): The HTTP verb. Includes DELETE, GET, HEAD, POST*, PUT.
+            Signed URLs do not support POST requests, except when working with
+            resumable uploads.
+        expires (int): The amount of time (in seconds) before the signed url will expire.
+        extension_headers (dict, optional): Optional request headers.
+        canonical_query_params(dict, optional): Optional query strings to add to the request
+        service_account_creds (dict, optional): JSON keyfile dict for Google
+            Service Account (can be obtained by calling `get_access_key`)
+        requester_pays_user_project (optional): User's Google project for billing
+
+    Returns:
+        str: Completed signed URL
+    """
+    service_account_creds = (
+        service_account_creds or config.GOOGLE_APPLICATION_CREDENTIALS
+    )
+    creds = service_account.Credentials.from_service_account_info(service_account_creds)
+
+    bucket_name = path_to_resource.split("/")[0]
+    object_name = "/".join(path_to_resource.split("/")[1:])
+    escaped_object_name = quote(object_name.encode(), safe=b"/~")
+    canonical_uri = "/{}".format(escaped_object_name)
+
+    datetime_now = datetime.datetime.utcnow()
+    request_timestamp = datetime_now.strftime("%Y%m%dT%H%M%SZ")
+    datestamp = datetime_now.strftime("%Y%m%d")
+
+    client_email = creds.service_account_email
+    credential_scope = "{}/auto/storage/goog4_request".format(datestamp)
+    credential = "{}/{}".format(client_email, credential_scope)
+
+    if extension_headers is None:
+        extension_headers = dict()
+
+    host = (
+        "{}.storage.googleapis.com".format(bucket_name)
+        if bucket_name
+        else "storage.googleapis.com"
+    )
+    extension_headers["host"] = host
+
+    canonical_headers = ""
+    ordered_headers = collections.OrderedDict(sorted(extension_headers.items()))
+
+    for k, v in ordered_headers.items():
+        lower_k = str(k).lower()
+        strip_v = str(v).lower()
+        canonical_headers += "{}:{}\n".format(lower_k, strip_v)
+
+    signed_headers = ";".join([str(k).lower() for k in ordered_headers])
+
+    if canonical_query_params is None:
+        canonical_query_params = dict()
+
+    if requester_pays_user_project is not None:
+        canonical_query_params["userProject"] = requester_pays_user_project
+
+    canonical_query_params["x-goog-algorithm"] = "GOOG4-RSA-SHA256"
+    canonical_query_params["x-goog-credential"] = credential
+    canonical_query_params["x-goog-date"] = request_timestamp
+    canonical_query_params["x-goog-expires"] = expires
+    canonical_query_params["x-goog-signedheaders"] = signed_headers
+
+    # sort params for deterministic hashing
+    ordered_query_parameters = collections.OrderedDict(
+        sorted(canonical_query_params.items())
+    )
```

Comment:
> we could change the case back just to match Google's sample code

> this would alter the way they are sorted and will therefore not work with requestor pays

> if the order of the headers matters, it should be explicit in the code, not implicit with the sorting. Let's add `userProject` at the right spot after sorting the other headers, and add a comment?

same here

add a comment saying "userProject" must be first (or whatever the requirement is), and why if you know why.
if it needs to be first, i would add it to `ordered_query_parameters` _after_ sorting the order params so that it is explicit. if it's just about ordering alphabetically in a case-insensitive way, you can comment that

Reply:
It's about sorting and is _case-sensitive_ . The reason why is already given in the comment.